media/main.js

```diff
@@ -0,0 +1,144 @@
+    const inputField = document.querySelector('#inputField');
+
+    inputForm.addEventListener('submit', (e) => handleSubmit(e, inputField));
+}
+
+function handleSubmit(e, inputField) {
+    e.preventDefault();
+    const message = inputField.value;
+    if (message.trim()) {
+        addMessage(message, 'User');
+        vscode.postMessage({
+            command: 'assistantRequest',
+            text: message
+        });
+        updateSubmitButton(true);
+    }
+    inputField.value = '';
+    addPendingMessage();
+}
+
+function handleAssistantResponse(completionResult) {
+    if (completionResult) {
+        addMessage(completionResult, 'Assistant');
+    } else {
+        addMessage('No response from Assistant', 'Assistant');
+    }
+}
+
+function updateSubmitButton(disabled) {
+    const submitButton = document.querySelector('#submitButton');
+    submitButton.disabled = disabled;
+    submitButton.innerHTML = disabled ? "..." : "Send";
+}
+
+function addPendingMessage() {
+    const messageElement = createMessageElement("Assistant", true);
+    appendMessageElement(messageElement);
+}
+
+function addPendingMessage() {
+    const messageElement = document.createElement('div');
+    messageElement.setAttribute('data-username', "Assistant");
+    messageElement.classList.add('message');
+    messageElement.classList.add('pending');
+    messageElement.innerHTML = `
+        <div class="loader">
+            <div class="dot"></div>
+            <div class="dot"></div>
+            <div class="dot"></div>
+        </div>
+    `;
+    chatbox.appendChild(messageElement);
+    chatbox.scrollTop = chatbox.scrollHeight;
+}
+
+function removePendingMessage() {
+    const pendingMessage = document.querySelector('.message.pending');
+    if (pendingMessage) {
+        pendingMessage.remove();
+    }
+}
+
+function addMessage(markdownText, sender) {
+    removePendingMessage();
+    const messageElement = createMessageElement(sender);
+    messageElement.innerHTML = marked.parse(markdownText);
+    appendMessageElement(messageElement);
+    renderMarkdown();
+    addCopyButtonToCodeBlocks();
+}
+
+function createMessageElement(sender) {
+    const messageElement = document.createElement('div');
+    messageElement.classList.add('message');
+    messageElement.setAttribute('data-username', sender);
+    return messageElement;
+}
+
+function appendMessageElement(messageElement) {
+    const chatbox = document.querySelector('#chatbox');
+    chatbox.appendChild(messageElement);
+    chatbox.scrollTop = chatbox.scrollHeight;
+}
+
+function renderMarkdown() {
+    marked.setOptions({
+        highlight: function(code, lang) {
+            const language = hljs.getLanguage(lang) ? lang : 'plaintext';
+            return hljs.highlight(code, { language }).value;
+        }
+    });
+    document.querySelectorAll('.message pre code').forEach((block) => {
+        hljs.highlightBlock(block);
+    });
+}
+
+function addCopyButtonToCodeBlocks() {
```

Comment:
This is pretty neat.

I was thinking in the long run we could also have "Apply" button. We know to which part of the original code the response corresponds to, so we could merge it back with a single button press.

Reply:
Exactly, I am thinking of this feature as well, even more we can add, like after click "Apply", we modify the source code with a `diff view`, and pop up two "Accept", "Discard" button in the editor.